prow/tide/history/history.go

```diff
@@ -53,30 +134,53 @@ type Record struct {
 }
 
 // New creates a new History struct with the specificed recordLog size limit.
-func New(maxRecordsPerKey int) *History {
-	return &History{
-		logs:         make(map[string]*recordLog),
+func New(maxRecordsPerKey int, gcsObject *storage.ObjectHandle) (*History, error) {
+	hist := &History{
+		logs:         map[string]*recordLog{},
 		logSizeLimit: maxRecordsPerKey,
+		gcsObj:       gcsObject,
+	}
+
+	if gcsObject != nil {
+		// Load existing history from GCS.
+		var err error
+		start := time.Now()
+		hist.logs, err = readHistory(maxRecordsPerKey, gcsStorageObject{gcsObject})
+		if err != nil {
+			return nil, fmt.Errorf("error loading history from GCS: %v", err)
+		}
+		logrus.WithField("duration", time.Since(start).String()).Debugf(
+			"Successfully read action history for %d pools.",
+			len(hist.logs),
+		)
 	}
+
+	return hist, nil
 }
 
 // Record appends an entry to the recordlog specified by the poolKey.
 func (h *History) Record(poolKey, action, baseSHA, err string, targets []prowapi.Pull) {
 	t := now()
 	sort.Sort(ByNum(targets))
+	h.addRecord(
+		poolKey,
+		&Record{
```

Comment:
If we don't use pointers we at least double the memory footprint of the action history. Is there some downside to using pointers here?

Reply:
Not clear to me how it would double memory. Seems like the pointer is created here and immediately shipped off to h.logs[poolKey].add(rec). Sending a non-pointer seems equivalent.

Reply:
The same Record is also stored in the recordLog's cached in-order slice and if we loaded from GCS there could even be a third copy.